File: aoc2020py/day1.py
import itertools

with open('input/1.txt') as f:
    expenses = []
    for line in f:
        expenses.append(int(line))

print(type(expenses[0]))
print(len(expenses))
print(sum(expenses))
print(len(list(itertools.permutations(expenses, 2))))

for t in itertools.combinations(expenses, 2):
    if t[0] + t[1] == 2020:
        x, y = t
        print(t)
print("Part 1", x * y)

for t in itertools.combinations(expenses, 3):
    if t[0] + t[1] + t[2] == 2020:
        x, y, z = t
        print(t)
print("Part 2", x * y * z)

File: adventofcode/2024js/day1.ts
function parseInput(file: string) {
  const left = [];
  const right = [];

  const lines = Deno.readTextFileSync(file);
  for (const line of lines.split("\n")) {
    const split = line.split(/\s+/);
    left.push(parseInt(split[0]));
    right.push(parseInt(split[1]));
  }

  return {
    left: left.sort(),
    right: right.sort(),
  };
}

// Part 1

function totalDistance(left: number[], right: number[]) {
  let sum = 0;
  for (let i = 0; i < left.length; i++) {
    sum += Math.abs(left[i] - right[i]);
  }
  return sum;
}

// Part 2

function similarityScore(left: number[], right: number[]) {
  const frequencies = new Map();
  for (const x of right) {
    frequencies.set(x, frequencies.get(x) + 1 || 1);
  }

  let sum = 0;
  for (let i = 0; i < left.length; i++) {
    sum += left[i] * frequencies.get(left[i]) || 0;
  }
  return sum;
}

const { left: l1, right: r1 } = parseInput("day1_sample.txt");
console.log(totalDistance(l1, r1));
console.log(similarityScore(l1, r1));

const { left: l2, right: r2 } = parseInput("day1_in.txt");
console.log(totalDistance(l2, r2));
console.log(similarityScore(l2, r2));
